### PR TITLE
Make ShufflerDataPipe deterministic for persistent DL and distributed DL

### LINKED ISSUE #426: Follow up on Shuffler interacted with DataLoader
### 🚀 The feature

After https://github.com/pytorch/pytorch/pull/77741 and https://github.com/pytorch/pytorch/pull/77855 are landed, there are two follow-up discussions on the shuffling behavior of `DataLoader` and `IterDataPipe`.
- For Distributed training, we currently have to let users to call `torch.manual_seed()` with the same seed across distributed processes to guarantee all shuffles have the same effect for `DataPipe`. However, for `Dataset`, users can provide the seed to`DistributedSampler` to control shuffle order. Side note: the seeding strategy for `DistributedSampler` is also bad as `seed + epoch`. For `IterableDataset`, we don't have the functionality to shuffle it in either distirbuted or mp environment.
  - Potential solution for `DataPipe`: We could check the environment whether or not in distributed setting (`torch.distributed. is_initialized()`. And, We could transmit seeds from the `DataLoader` in Rank 0 to other `DataLloader`s. And each `DataLoader` in distributed process needs to wait until this environment variable is set, then it should be able to set the same seed for all `Shuffler`s.
- For persistent workers, , each persistent worker process would set the same seed to `Shuffler`s deterministically. However, users won't be able to reset seeds for persistent workers after the first iteration starts via `torch.manual_seed` (The `seed` and RNG are persistent in worker process). But, for `Dataset`, users still can reset seed in the main process via `Sampler`.
  - Potential solution: Use [`_ResumeIteration`](https://github.com/pytorch/pytorch/blob/97fa1d317f68fec5c99eb4144d01795d3fb2a9e1/torch/utils/data/_utils/worker.py#L117-L120) to pass seed from main process to persistent process.

### Motivation, pitch

- Make `Shuffler` automatically seeded in distributed training
- Make `Shuffler` can be seeded per epoch even with persistent workers

### Alternatives

_No response_

### Additional context

cc: @NicolasHug @VitalyFedyunin

### PR DESCRIPTION
**Pull request by ejguan**
Fixes https://github.com/pytorch/data/issues/426

This PR introduces two main changes:
- It ensures the `ShufflerDataPipe` would share the same seed across distributed processes.
- Users can reset `shuffle` for persistent workers per epoch.

Detail:
- `shared_seed` is shared across distributed and worker processes. It will seed a `shared_rng` to provide seeds to each `ShufflerDataPipe` in the pipeline
- `worker_loop` now accepts a new argument of `shared_seed` to accept this shared seed.
- The `shared_seed` is attached to `_ResumeIteration` for resetting seed per epoch for `persistent worker`
- I choose not to touch `base_seed` simply for BC issue

I used this [script](https://gist.github.com/ejguan/d88f75fa822cb696ab1bc5bc25844f47) to test the result with `world_size=4`. Please check the result in: https://gist.github.com/ejguan/6ee2d2de12ca57f9eb4b97ef5a0e300b

You can see there isn't any duplicated/missing element for each epoch. And, with the same seed, the order of data remains the same across epochs.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -521,12 +522,36 @@ def _create_warning_msg(num_worker_suggest, num_worker_created, cpuset_checked):
...
+    def _get_shared_seed(self):
+        if isinstance(self.dataset, IterDataPipe):
+            _shared_tensor_seed = torch.empty((), dtype=torch.int64).random_(generator=self.generator)
+            if dist.is_available() and dist.is_initialized():
+                rank = dist.get_rank()
+                if rank == 0:
+                    ws = dist.get_world_size()
+                    reqs = []
+                    for rank_idx in range(1, ws):
+                        req = dist.isend(tensor=_shared_tensor_seed, dst=rank_idx)
```
Comment (VitalyFedyunin) [inline torch/utils/data/dataloader.py:534]: Feels like you need to `tag = 'dataloader_shared_seed'` here

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -521,12 +522,36 @@ def _create_warning_msg(num_worker_suggest, num_worker_created, cpuset_checked):
...
+                        reqs.append(req)
+                    for req in reqs:
+                        req.wait()
+                else:
+                    dist.recv(tensor=_shared_tensor_seed, src=0)
```
Comment (VitalyFedyunin) [inline torch/utils/data/dataloader.py:539]: I feel that having tagged `isend` and `recv` is safer.

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:534]: It totally makes sense. Since `tag` only accepts integer, I will use `rank` as the tag to transmit `_shared_tensor_seed`

Comment (VitalyFedyunin) [reply torch/utils/data/dataloader.py:534]: It looks like an unused feature inside of the `dist` code. As for me `tag` should prevent other parts of the code from receiving this tensor. `rank_id` is fine for now, but It is better to talk with dist team, to figure out if we need to define tag constants somewhere and have something like `_DIST_TAG_DOWNLOADER_ = 1000`

**Review by VitalyFedyunin — APPROVED** [review]

File: torch/utils/data/_utils/worker.py [hunk 3]
```diff
@@ -201,7 +201,7 @@ def mix(x, y):
 
 def _worker_loop(dataset_kind, dataset, index_queue, data_queue, done_event,
                  auto_collation, collate_fn, drop_last, base_seed, init_fn, worker_id,
-                 num_workers, persistent_workers):
+                 num_workers, persistent_workers, shared_seed):
```
Comment (NivekT) [inline torch/utils/data/_utils/worker.py:204]: nit: it seems like the argument `persistent_workers` isn't used? Do we still need it?

Comment (ejguan) [reply torch/utils/data/_utils/worker.py:204]: I actually don't know. Probably not needed if not reference in this `worker_loop` function. Even if we want to remove it, I won't keep the change in this PR.

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/dataloader.py:534]: Reasonable to me!

**Comment by janeyx99** [comment]
Link: https://hud.pytorch.org/minihud#b769a0e18b2d87884e1db8dfca2010e22494de69
